torch/csrc/jit/codegen/cuda/test/test_gpu_outer_reduction.cpp

```diff
@@ -221,6 +219,1125 @@ TEST_F(NVFuserTest, FusionGroupedGridWelfordOuterOpt_CUDA) {
+      use_compute_with = atoi(getenv("COMPUTE_WITH")) != 0;
+    }
+  }
+};
+
+void scheduleNormalization(Fusion& fusion, const OuterReductionParams& params) {
+  // Cache inputs
+  std::vector<TensorView*> input_caches;
+  for (auto input_tv : ir_utils::filterByType<TensorView>(fusion.inputs())) {
+    // Skip unused inputs
+    if (input_tv->uses().empty()) {
+      continue;
+    }
+
+    // Assumes any 1D tensor is just a Float or Half tensor that can be reused
+    // without recomputation
+    if (input_tv->nDims() == 1) {
+      input_caches.push_back(input_tv->cacheAfter());
+      continue;
+    }
+
+    // Any other tensor is assumed to be a 4D tensor that requires recomputation
+    TORCH_CHECK(
+        input_tv->nDims() == 4,
+        "Unexpected input tensor: ",
+        input_tv->toString());
+    // If the input type is Half, it must be cast to Float
+    if (input_tv->getDataType() == DataType::Half) {
+      TORCH_CHECK(
+          input_tv->uses().size() == 1,
+          "Unexpected input tensor: ",
+          input_tv->toString());
+      auto cast_expr = dynamic_cast<UnaryOp*>(input_tv->uses().at(0));
+      TORCH_CHECK(
+          cast_expr != nullptr &&
+              cast_expr->getUnaryOpType() == UnaryOpType::Cast,
+          "Unexpected input tensor: ",
+          input_tv->toString());
+      auto cast_tv = dynamic_cast<TensorView*>(cast_expr->out());
+      TORCH_CHECK(cast_tv != nullptr);
+      for (auto use : cast_tv->uses()) {
+        auto replica = RecomputeTv::recompute(cast_tv);
+        ir_utils::replaceValInExpr(use, cast_tv, replica);
+      }
+    }
+    input_caches.push_back(input_tv->cacheAfter());
+  }
+
+  // Cache outputs
+  std::vector<TensorView*> output_caches;
+  for (auto output_tv : ir_utils::filterByType<TensorView>(fusion.outputs())) {
+    output_caches.push_back(output_tv->cacheBefore());
+  }
+
+  // Find reduction TVs
+  const auto used_math_vals = fusion.usedMathVals();
+  std::vector<Expr*> reduction_exprs;
+  std::vector<TensorView*> reduction_tvs;
+
+  for (auto expr : fusion.exprs()) {
+    if (ir_utils::isReductionTvOp(expr)) {
+      reduction_exprs.push_back(expr);
+      reduction_tvs.push_back(ir_utils::getTvOutput(expr));
+    }
+  }
+
+  TORCH_CHECK(!reduction_exprs.empty(), "No reduction found");
+
+  // Apply horizontal grouping before rfactor
+  if (reduction_exprs.size() > 1) {
+    groupReductions(reduction_tvs);
+  }
+
+  TORCH_CHECK(
+      dataTypeSize(DataType::Half) * params.vec * reduction_tvs.size() <= 16,
+      "Invalid vectorization");
+
+  auto reduction_tv = reduction_tvs.at(0);
+
+  // Transform a reduction TV as the reference
+  reduction_tv->merge(0)->merge(0);
+
+  // [r(I0), i(I1)]
+  reduction_tv->split(0, params.tidy);
+  // [r/(I0/TIDy), r(TIDy), i(I1)]
+  reduction_tv->axis(1)->parallelize(ParallelType::TIDy);
+
+  // When outer_pb is true, use the outer split for the persistent
+  // buffer domain
+  int unswitch_pos = 0;
+  if (params.outer_pb) {
```

Comment:
Outer: `[r(PB), blocckIdx.y(r/(I0/TIDy/PB)), r(TIDy), i(I1)]`
Inner: `[blockIdx.y(r/(I0/TIDy/PB)), r(PB), r(TIDy), i(I1)]`

Since the thread blocks are more or less synchronously executed, I wondered if the blockIdx.y-parallelized dimension should be placed inner than the persistent-buffer dimension. That might require less TLB entries, for example. I didn't see any better result with the outer option, though.